components/base/kinematicbase/kinematics.go

```diff
@@ -102,50 +107,91 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
 
-	// this loop polls the error state and issues a corresponding command to move the base to the objective
-	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	cancelContext, cancel := context.WithCancel(ctx)
 
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+	utils.PanicCapturingGo(func() {
+		// this loop polls the error state and issues a corresponding command to move the base to the objective
+		// when the base is within the positional threshold of the goal, exit the loop
+		for contextErr := ctx.Err(); contextErr == nil; contextErr = ctx.Err() {
+			utils.SelectContextOrWait(ctx, 100*time.Millisecond)
+			col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
+			if err != nil {
+				movementErr <- err
+			}
+			if !col {
+				movementErr <- errors.New("base has deviated too far from path")
+			}
+
+			// get to the x, y location first - note that from the base's perspective +y is forward
+			desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
+			commanded, err := ddk.issueCommand(cancelContext, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
+			if err != nil {
+				movementErr <- err
+			}
 
-		if !commanded {
-			// no command to move to the x, y location was issued, correct the heading and then exit
-			if commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{current[0], current[1], desired[2]}); err == nil {
-				if !commanded {
-					return nil
+			if !commanded {
+				// no command to move to the x, y location was issued, correct the heading and then exit
+				if commanded, err := ddk.issueCommand(cancelContext, current, []referenceframe.Input{current[0], current[1], desired[2]}); err == nil {
+					if !commanded {
+						movementErr <- nil
+					}
+				} else {
+					movementErr <- err
 				}
-			} else {
-				return err
+			}
+
+			current, err = ddk.CurrentInputs(cancelContext)
+			if err != nil {
+				movementErr <- err
 			}
 		}
+		movementErr <- ctx.Err()
+	})
 
-		current, err = ddk.CurrentInputs(ctx)
-		if err != nil {
-			return err
+	// goroutine for watching for movement timeout
+	utils.PanicCapturingGo(func() {
+		lastUpdate := time.Now()
+		prevDistErr := math.Inf(-1)
+		prevHeadingErr := math.Inf(-1)
+
+		for {
+			currentPose, err := ddk.CurrentInputs(ctx)
```

Comment:
This doesn't return a pose so preference to rename to just `current`

Reply:
can't name it `current` because the other goroutine is modifying the `current` variable that was created at the top of the function. but will rename to smth else